### PR TITLE
Add GetCompatibilePackageTargetFrameworks  Task to the new packageValidation package

### PR DESCRIPTION
**Pull request by Anipik**
GetCompatibilePackageTargetFrameworks tasks returns the list of tfms where a particular package should be installed and closure tested
we determine the compatible tfms by already defined compatible rules in the nuget and also prune the list by removing some older frameworks.
Added the tests as well.

### CODE REVIEW

File: src/Microsoft.DotNet.PackageValidation.Tests/Microsoft.DotNet.PackageValidation.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="15.7.179" />
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation.Tests/Microsoft.DotNet.PackageValidation.Tests.csproj:9]: Why a hard-coded version instead of one from Versions.props?

File: src/Microsoft.DotNet.PackageValidation/Asset.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using NuGet.Frameworks;
+
+namespace Microsoft.DotNet.PackageValidation
+{
+    public class Asset
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation/Asset.cs:8]: Asset is probably not descriptive enough and might clash with other types in future. Consider using `PackageAsset` and `PackageAssetType` instead.

File: src/Microsoft.DotNet.PackageValidation/Asset.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        public NuGetFramework TargetFramework { get; set; }
+        public string Rid { get; set; }
+        public string PackagePath { get; set; }
+        public AssetType AssetType { get; set; }
+
+        public Asset(NuGetFramework targetFramework, string runtimeIdentifier, string packagePath1, AssetType assetType)
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation/Asset.cs:15]: Why packagePath1?

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Diagnostics;
+
+namespace Microsoft.DotNet.PackageValidation
+{
+    public class GetCompatibilePackageTargetFrameworks : BuildTask
+    {
+        private static List<NuGetFramework> allTargetFrameworks = allTargetFrameworks = new List<NuGetFramework>();
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs:17]:
```suggestion
        private static List<NuGetFramework> allTargetFrameworks = new();
```

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        private static Dictionary<NuGetFramework, HashSet<NuGetFramework>> packageTfmMapping = new Dictionary<NuGetFramework, HashSet<NuGetFramework>>();
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs:18]:
```suggestion
        private static Dictionary<NuGetFramework, HashSet<NuGetFramework>> packageTfmMapping = new();
```

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+        [Required]
+        public string[] PackagePaths { get; set; }
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs:21]: I don't know too much about string array vs ITaskItem array, but usually any array I saw in task uses ITaskItem. Why not here?

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard11);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard12);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard13);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard14);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard15);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard16);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard17);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard20);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard21);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.UAP10);
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs:99]: Static known frameworks shouldn't be hardcoded in a compiled asset and instead be passed into the task and supplied via the targets file that invokes the task.

File: src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<Project>
+  <PropertyGroup>
+    <PackageValidationDllPath>$(MSBuildThisFileDirectory)\..\tools\net5.0\Microsoft.DotNet.PackageValidation.dll</PackageValidationDllPath>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props:3]:
```suggestion
    <DotNetPackageValidationAssembly>$(MSBuildThisFileDirectory)\..\tools\net5.0\Microsoft.DotNet.PackageValidation.dll</DotNetPackageValidationAssembly>
```

to be consistent with other places like the TFM SKD: https://github.com/dotnet/arcade/blob/6d7ec926c8ae962ec5c1f5beeef3f67f9eeda635/src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.props#L11-L12.

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs (hunk 6 above)
Comment (Anipik) [reply src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs:21]: AFAIK the taskitem should be used when are adding or doing any operations on item metadata, which is not the case with the input here

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs (hunk 7 above)
Comment (Anipik) [reply src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs:99]: why would it be better to pass the static list, thats not gonna be changed, be provided as an input ?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GetComaptiblePackageTargetFrameworks.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (safern) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GetComaptiblePackageTargetFrameworks.cs:1]: The file name has a typo: `GetComaptiblePackageTargetFrameworks` -> `GetCompatiblePackageTargetFrameworks`

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+using Microsoft.Build.Utilities;
+using Microsoft.DotNet.Build.Tasks;
+using NuGet.Frameworks;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.DotNet.PackageValidation
+{
+    public class GetCompatibilePackageTargetFrameworks : BuildTask
```
Comment (safern) [inline src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs:14]:
```suggestion
    public class GetCompatiblePackageTargetFrameworks : BuildTask
```

Also the file name.

File: src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<Project>
```
Comment (safern) [inline src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props:1]: Why not include these tasks as part of the `Packaging` tasks package?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GetComaptiblePackageTargetFrameworks.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+                Package package = NupkgParser.CreatePackageObject(PackagePath);
+                var packageTargetFrameworks = package.PackageAssets.Where(t => t.AssetType != AssetType.RuntimeAsset).Select(t => t.TargetFramework).Distinct();
+                CompatibilityTable table = new CompatibilityTable(packageTargetFrameworks);
+                List<string> frameworksToTest = new List<string>();
+
+                foreach (var item in TargetFrameworks)
+                {
+                    IEnumerable<NuGetFramework> compatible = null;
+                    table.TryGetCompatible(NuGetFramework.Parse(item), out compatible);
+                    if (compatible != null)
```
Comment (safern) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GetComaptiblePackageTargetFrameworks.cs:37]: Can this be rewritten as?

```suggestion
                    if (table.TryGetCompatible(NuGetFramework.Parse(item), out IEnumerable<NuGetFramework> compatible))
```

**Comment by missymessa** [comment]
We have an abstract class in Arcade to facilitate in DI for the MSBuild BuildTask classes: https://github.com/dotnet/arcade/blob/master/Documentation/Mechanics/MSBuildTaskDependencyInjection.md

File: src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props (hunk 11 above)
Comment (Anipik) [reply src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props:1]: 95% of the stuff in packaging tasks package will not be obselete after package validation & porting projects to the pack task. Most of the code will be entirely different from the packaging code.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GetComaptiblePackageTargetFrameworks.cs (hunk 12 above)
Comment (Anipik) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GetComaptiblePackageTargetFrameworks.cs]: this file is redundant & has been deleted.

File: src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props (hunk 11 above)
Comment (safern) [reply src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props:1]: Where will the other 5% in the packaging tasks will go?

File: src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs (hunk 7 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs]: In case we would ever want to mutate these items, we could do so via msbuild without the need to rebuild the assembly and flow it into dotnet/runtime or any other consumer repo.

Comment (ViktorHofer) [reply src/Microsoft.DotNet.PackageValidation/GetCompatibilePackageTargetFrameworks.cs]: Ie wouldn't we want to remove the UAP entry there when we stop supporting UAP entirely?

File: src/Microsoft.DotNet.PackageValidation.Tests/Microsoft.DotNet.PackageValidation.Tests.csproj [hunk 13]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <XUnitPublishTargetFramework>net5.0</XUnitPublishTargetFramework>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation.Tests/Microsoft.DotNet.PackageValidation.Tests.csproj:5]: Curious, why is this necessary?

File: src/Microsoft.DotNet.PackageValidation.Tests/Microsoft.DotNet.PackageValidation.Tests.csproj [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" />
+    <PackageReference Include="NuGet.Frameworks" Version="$(NuGetVersion)" />
+
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.PackageValidation.Tests/Microsoft.DotNet.PackageValidation.Tests.csproj:12]: nit: empty line

File: src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props (hunk 11 above)
Comment (Anipik) [reply src/Microsoft.DotNet.PackageValidation/build/Microsoft.DotNet.PackageValidation.props:1]: probably in the validation package or to the runtime repo depending on the tasks. The only tasks currently in that bucket is the verify closure one.

**Review by ViktorHofer — APPROVED** [review]
I think this is a reasonable point to start. I think the list of hardcoded frameworks shouldn't be part of the compiled assembly but I might be missing context and there's actually a reason for not having it exposed in msbuild. Besides that, LGTM. Thanks Anirudh

**Comment by Anipik** [comment]
> I think the list of hardcoded frameworks shouldn't be part of the compiled assembly but I might be missing context and there's actually a reason for not having it exposed in msbuild

The main reason is to have minimal inputs that a user has to specify while consuming in this package. I think we can make a final decision later when we know more about how this package is going to be consumed(part of sdk or standalone package) and how this fits as a consumer scenario.

Because dropping the tfm from the list might only be relevant to the runtime scenario where we run this to verify that there are no duplicated types, hence customer might not have to worry about consuming the latest package.

**Comment by ViktorHofer** [comment]
> The main reason is to have minimal inputs that a user has to specify while consuming in this package. I think we can make a final decision later when we know more about how this package is going to be consumed(part of sdk or standalone package) and how this fits as a consumer scenario.

That doesn't hinder you from defining the static list via msbuild items in the package's props file which is auto-imported when referencing the package.

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+            }
+            catch (Exception e)
+            {
+                Log.LogErrorFromException(e, showStackTrace: false);
+            }
+            
+            return result && !Log.HasLoggedErrors;
+        }
+
+        public static List<NuGetFramework> GetTestFrameworks(List<NuGetFramework> packageTargetFrameworks)
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:53]: nit: this can be IEnumerable and you can avoid calls ToList

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+                    packageTfmMapping[forwardTfm].Add(reverseTfm);
+                }
+                else
+                {
+                    packageTfmMapping.Add(forwardTfm, new HashSet<NuGetFramework> { reverseTfm });
+                }
+            }
+        }
+    
+        public List<ITaskItem> CreateItemFromTestFramework(string title, string version, List<NuGetFramework> testFrameworks, string rids)
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:116]: Similarly List vs IEnumerable.  Try to avoid using concrete collections unless you actually need to.  It allocates unnecessary arrays.

General rules:
If all you care about is enumeration use IEnumerable
If you need to enumerate many times, but don't need to modify results, consider an array.
If you need to modify the collection then use IList.

File: src/Microsoft.DotNet.PackageValidation/NupkgParser.cs [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            string title = nuspecReader.GetTitle();
+            string version = nuspecReader.GetVersion().ToString();
+            IEnumerable<PackageDependencyGroup> dependencyGroups = nuspecReader.GetDependencyGroups();
+
+            foreach (var item in dependencyGroups)
+            {
+                packageDependencies.Add(item.TargetFramework, item.Packages.ToList());
+            }
+
+            var files = nupkgReader.GetFiles().ToList().Where(t => t.EndsWith(".dll")).Where(t => t.Contains(title + ".dll"));
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/NupkgParser.cs:31]: The call `ToList` here is allocating a List that's never used anywhere since you immediately filter it.

Also your where statement seems to be redundant.  Can it just be `t.EndsWith(title + ".dll", StringComparison.OrdinalIgnoreCase)`

File: src/Microsoft.DotNet.PackageValidation/NupkgParser.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+            foreach (var file in files)
+            {
+                packageAssets.Add(ExtractAssetFromFile(file));
+            }
+
+            return new Package(title, version, packageAssets, packageDependencies);
+        }
+
+        public static PackageAsset ExtractAssetFromFile(string filePath)
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/NupkgParser.cs:40]: This name should be changed. Nothing is being extracted here.  Consider naming `Get` rather than `Extract`

File: src/Microsoft.DotNet.PackageValidation/NupkgParser.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            PackageAsset asset = null;
+            if (filePath.StartsWith("ref"))
+            {
+                var stringParts = filePath.Split(@"/");
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/NupkgParser.cs:45]: pull the split out of these if statements.  Make sure you are doing length checks on the resultant array to avoid index out of range on packages with unexpected content (like a file named `ref` or a file under `runtimes\foo.txt` etc).  When doing string comparisons use OrdinalIgnoreCase.

File: src/Microsoft.DotNet.PackageValidation/Package.cs [hunk 20]
```diff
@@ -0,0 +1,24 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using System.Collections.Generic;
+
+namespace Microsoft.DotNet.PackageValidation
+{
+    public class Package
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/Package.cs:10]: Rather than creating our own content model for the package, did you consider using NuGet's types?   This would then leverage NuGet's asset discovery logic rather than us writing our own.

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs

See FindItemGroups.

Here's an example of its use:
https://github.com/dotnet/arcade/blob/7f13798e5f567b72ffe63205bf49839245f0f8c1/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetAssetResolver.cs#L280

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard11);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard12);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard13);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard14);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard15);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard16);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard17);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard20);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.NetStandard21);
+            allTargetFrameworks.Add(FrameworkConstants.CommonFrameworks.UAP10);
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:98]: Agree with what others have said about this, better to have this data in props.  If it were easy to calculate it from NuGet that would be even better, but we found that wasn't the case so better to just define in props so it can more easily be controlled by caller.

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+    public class GetCompatiblePackageTargetFrameworks : BuildTask
+    {
+        private static List<NuGetFramework> allTargetFrameworks = allTargetFrameworks = new();
+        private static Dictionary<NuGetFramework, HashSet<NuGetFramework>> packageTfmMapping = new();
+
+        [Required]
+        public string[] PackagePaths { get; set; }
+
+        [Output]
+        public ITaskItem[] TestProjects { get; set; }
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:23]: Try to document your inputs and outputs.  What's the expected identity and metadata and what is it used for.  This is the tasks's public API.

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+                }
+            }
+        }
+    
+        public List<ITaskItem> CreateItemFromTestFramework(string title, string version, List<NuGetFramework> testFrameworks, string rids)
+        {
+            List<ITaskItem> testprojects = new List<ITaskItem>();
+            foreach (var framework in testFrameworks)
+            {
+                var supportedPackage = new TaskItem(title);
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:121]: title here is confusing.  Shouldn't this be packageId?

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs (hunk 22 above)
Comment (ericstj) [reply src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:23]: Naming here is off, consider TestPackages

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+            List<string> rids = new List<string>();
+            foreach (var item in package.PackageAssets)
+            {
+                if (item.AssetType == AssetType.RuntimeAsset)
+                {
+                    if (!rids.Contains(item.Rid + "-x64"))
+                        rids.Add(item.Rid + "-x64");
+                }
+            }
+            return string.Join(";", rids);
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:147]: Why are you joining here when other methods deal with lists/enumerables?  Consider moving the join to the CreateItemFromTestFramework method where this is used.

File: src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+                }
+                testprojects.Add(supportedPackage);
+            }
+
+            return testprojects;
+        }
+    
+        public string GetRidsFromPackage(Package package)
+        {
+            List<string> rids = new List<string>();
```
Comment (ericstj) [inline src/Microsoft.DotNet.PackageValidation/GetCompatiblePackageTargetFrameworks.cs:138]: You can avoid a list allocation here by making this method return an enumerable and yield returning the items

**Review by ericstj — COMMENTED** [review]
Sorry for the late feedback

**Comment by ViktorHofer** [comment]
@Anipik are you planning to follow-up on the feedback provided by @ericstj and me (less allocation and no static list in code)?

**Comment by Anipik** [comment]
> @Anipik are you planning to follow-up on the feedback provided by @ericstj and me (less allocation and no static list in code)?

yes i will try to get it later today.
